[BUG] Batched silhouette score has a stream-ordering race

**Describe the bug**

`cuvs::stats::silhouette_score_batched` intermittently returns a different result from `cuvs::stats::silhouette_score` when the RAFT handle has a CUDA stream pool.

The mismatch becomes observable under concurrent GPU load. It affects multiple distance metrics, suggesting that the problem is in the batched orchestration rather than a metric-specific implementation.

**Steps/Code to reproduce bug**

The following program compares the batched and non-batched APIs using identical deterministic input:

```cpp
#include <cuvs/stats/silhouette_score.hpp>
#include <raft/core/device_mdarray.hpp>
#include <raft/core/resource/cuda_stream.hpp>
#include <raft/core/resource/cuda_stream_pool.hpp>
#include <raft/util/cudart_utils.hpp>
#include <rmm/cuda_stream_pool.hpp>

#include <cmath>
#include <iostream>
#include <memory>
#include <vector>

int main()
{
  constexpr int64_t n_rows = 4096;
  constexpr int64_t n_cols = 2;

  std::vector<float> X(n_rows * n_cols);
  std::vector<int> labels(n_rows);

  for (int64_t i = 0; i < n_rows; ++i) {
    X[2 * i]     = std::sin(0.01f * i);
    X[2 * i + 1] = std::cos(0.013f * i);
    labels[i]    = i % 2;
  }

  raft::resources handle;
  raft::resource::set_cuda_stream_pool(
    handle, std::make_shared<rmm::cuda_stream_pool>(4));

  auto stream   = raft::resource::get_cuda_stream(handle);
  auto d_X      = raft::make_device_matrix<float, int64_t>(handle, n_rows, n_cols);
  auto d_labels = raft::make_device_vector<int, int64_t>(handle, n_rows);

  raft::update_device(d_X.data_handle(), X.data(), X.size(), stream);
  raft::update_device(d_labels.data_handle(), labels.data(), labels.size(), stream);

  auto X_view = raft::make_device_matrix_view<const float, int64_t>(
    d_X.data_handle(), n_rows, n_cols);
  auto labels_view = raft::make_device_vector_view<const int, int64_t>(
    d_labels.data_handle(), n_rows);

  constexpr auto metric =
    cuvs::distance::DistanceType::L2SqrtUnexpanded;

  float expected = cuvs::stats::silhouette_score(
    handle, X_view, labels_view, std::nullopt, 2, metric);
  float actual = cuvs::stats::silhouette_score_batched(
    handle, X_view, labels_view, std::nullopt, 2, n_rows, metric);

  std::cout << "non-batched: " << expected << '\n'
            << "batched:     " << actual << '\n'
            << "difference:  " << std::abs(actual - expected) << '\n';

  return std::abs(actual - expected) > 1e-4f;
}
```

Running this repeatedly with concurrent processes reproduces the mismatch intermittently. In one run with 512 attempts and 8 concurrent processes, 5 attempts failed:

```text
Total:        512
Passed:       507
Failed:       5
Failure rate: 0.98%
```

One observed mismatch was:

```text
non-batched: -0.000485941
batched:     -0.00263832
difference:   0.00215238
```

A control run without an explicit stream pool completed 256 attempts without a mismatch.

<details>
<summary>Full self-contained reproducing script</summary>

```bash
#!/usr/bin/env bash

set -euo pipefail

runs=${1:-512}
jobs=${2:-8}

tmp_dir=$(mktemp -d)
source_file="$tmp_dir/mre_silhouette_score_batched.cu"
binary="$tmp_dir/mre_silhouette_score_batched"
trap 'rm -rf "$tmp_dir"' EXIT

cat > "$source_file" <<'CU'
#include <cuvs/stats/silhouette_score.hpp>
#include <raft/core/device_mdarray.hpp>
#include <raft/core/resource/cuda_stream.hpp>
#include <raft/core/resource/cuda_stream_pool.hpp>
#include <raft/util/cudart_utils.hpp>
#include <rmm/cuda_stream_pool.hpp>

#include <cmath>
#include <iostream>
#include <memory>
#include <vector>

int main()
{
  constexpr int64_t n_rows = 4096;
  constexpr int64_t n_cols = 2;

  std::vector<float> X(n_rows * n_cols);
  std::vector<int> labels(n_rows);
  for (int64_t i = 0; i < n_rows; ++i) {
    X[2 * i]     = std::sin(0.01f * i);
    X[2 * i + 1] = std::cos(0.013f * i);
    labels[i]    = i % 2;
  }

  raft::resources handle;
  raft::resource::set_cuda_stream_pool(
    handle, std::make_shared<rmm::cuda_stream_pool>(4));
  auto stream   = raft::resource::get_cuda_stream(handle);
  auto d_X      = raft::make_device_matrix<float, int64_t>(handle, n_rows, n_cols);
  auto d_labels = raft::make_device_vector<int, int64_t>(handle, n_rows);
  raft::update_device(d_X.data_handle(), X.data(), X.size(), stream);
  raft::update_device(d_labels.data_handle(), labels.data(), labels.size(), stream);

  auto X_view = raft::make_device_matrix_view<const float, int64_t>(
    d_X.data_handle(), n_rows, n_cols);
  auto labels_view = raft::make_device_vector_view<const int, int64_t>(
    d_labels.data_handle(), n_rows);
  constexpr auto metric = cuvs::distance::DistanceType::L2SqrtUnexpanded;

  float expected = cuvs::stats::silhouette_score(
    handle, X_view, labels_view, std::nullopt, 2, metric);
  float actual = cuvs::stats::silhouette_score_batched(
    handle, X_view, labels_view, std::nullopt, 2, n_rows, metric);

  std::cout << "non-batched: " << expected << '\n'
            << "batched:     " << actual << '\n'
            << "difference:  " << std::abs(actual - expected) << '\n';
  return std::abs(actual - expected) > 1e-4f;
}
CU

"$CONDA_PREFIX/bin/nvcc" -std=c++20 \
  -I"$CONDA_PREFIX/include" \
  -L"$CONDA_PREFIX/lib" \
  -Xlinker -rpath -Xlinker "$CONDA_PREFIX/lib" \
  "$source_file" \
  -lcuvs -lraft -lrmm \
  -o "$binary"

# shellcheck disable=SC2016  # Expand positional parameters in each child shell.
seq "$runs" | xargs -P "$jobs" -n1 bash -c '
  output=$("$1" 2>&1) || printf "%s\n" "$output" > "$2/failure.$3"
' _ "$binary" "$tmp_dir"

shopt -s nullglob
failure_files=("$tmp_dir"/failure.*)
failed=${#failure_files[@]}
passed=$((runs - failed))

for failure_file in "${failure_files[@]}"; do
  printf '\nFailure in attempt %s:\n' "${failure_file##*.}"
  cat "$failure_file"
done

failure_rate=$(awk -v failed="$failed" -v runs="$runs" \
  'BEGIN { printf "%.2f", 100 * failed / runs }')

printf '\nSummary:\n'
printf '  Total:        %d\n' "$runs"
printf '  Passed:       %d\n' "$passed"
printf '  Failed:       %d\n' "$failed"
printf '  Failure rate: %s%%\n' "$failure_rate"

((failed == 0))
```

Run with:

```bash
./run_mre_silhouette_score_batched.sh 512 8
```

</details>

**Expected behavior**

`cuvs::stats::silhouette_score_batched` should return the same result as `cuvs::stats::silhouette_score`, within floating-point tolerance, regardless of whether the RAFT handle has a stream pool.

**Environment details**

- Environment location: bare metal
- Method of cuVS install: built from source
- Input type: `float32`
- Stream pool size: 4
- Concurrent processes: 8

**Additional context**

The leading hypothesis is missing stream ordering in `cpp/src/stats/detail/batched/silhouette_score.cuh`.

For each chunk, `get_pairwise_distance` allocates its output buffer on a selected pool stream, but calls `cuvs::distance::pairwise_distance` with the original RAFT handle. The distance computation therefore uses the handle's main stream. `compute_chunked_a_b` subsequently consumes the buffer on the selected pool stream.

The existing `wait_stream_pool_on_stream` call only orders main-stream work submitted before the batch loop. It does not establish a dependency between pairwise-distance work later submitted to the main stream and the consumer subsequently submitted to the pool stream.

This may allow `compute_chunked_a_b` to read a distance buffer before its producer has completed.

--- a/cpp/src/stats/detail/batched/silhouette_score.cuh
+++ b/cpp/src/stats/detail/batched/silhouette_score.cuh
@@ -1,5 +1,5 @@
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION.
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 
@@ -128,16 +128,13 @@ rmm::device_uvector<value_t> get_pairwise_distance(raft::resources const& handle
                                                    value_idx& n_left_rows,
                                                    value_idx& n_right_rows,
                                                    value_idx& n_cols,
-                                                   cuvs::distance::DistanceType metric,
-                                                   cudaStream_t stream)
+                                                   cuvs::distance::DistanceType metric)
 {
+  auto stream = raft::resource::get_cuda_stream(handle);
   rmm::device_uvector<value_t> distances(n_left_rows * n_right_rows, stream);
 
-  raft::resources stream_handle(handle);
-  raft::resource::set_cuda_stream(stream_handle, stream);
-
   cuvs::distance::pairwise_distance(
-    stream_handle,
+    handle,
     raft::make_device_matrix_view<const value_t, int64_t>(left_begin, n_left_rows, n_cols),
     raft::make_device_matrix_view<const value_t, int64_t>(right_begin, n_right_rows, n_cols),
     raft::make_device_matrix_view<value_t, int64_t>(distances.data(), n_left_rows, n_right_rows),
@@ -221,6 +218,8 @@ value_t silhouette_score(
       ++n_iters;
 
       auto chunk_stream = raft::resource::get_next_usable_stream(handle, i + chunk * j);
+      raft::resources chunk_handle(handle);
+      raft::resource::set_cuda_stream(chunk_handle, chunk_stream);
 
       const auto* left_begin  = X + (i * n_cols);
       const auto* right_begin = X + (j * n_cols);
@@ -229,7 +228,7 @@ value_t silhouette_score(
       auto n_right_rows = (j + chunk) < n_rows ? chunk : (n_rows - j);
 
       rmm::device_uvector<value_t> distances = get_pairwise_distance(
-        handle, left_begin, right_begin, n_left_rows, n_right_rows, n_cols, metric, chunk_stream);
+        chunk_handle, left_begin, right_begin, n_left_rows, n_right_rows, n_cols, metric);
 
       compute_chunked_a_b(handle,
                           a_ptr,

--- a/cpp/src/stats/detail/batched/silhouette_score.cuh
+++ b/cpp/src/stats/detail/batched/silhouette_score.cuh
@@ -133,8 +133,11 @@ rmm::device_uvector<value_t> get_pairwise_distance(raft::resources const& handle
 {
   rmm::device_uvector<value_t> distances(n_left_rows * n_right_rows, stream);
 
+  raft::resources stream_handle(handle);
+  raft::resource::set_cuda_stream(stream_handle, stream);
+
   cuvs::distance::pairwise_distance(
-    handle,
+    stream_handle,
     raft::make_device_matrix_view<const value_t, int64_t>(left_begin, n_left_rows, n_cols),
     raft::make_device_matrix_view<const value_t, int64_t>(right_begin, n_right_rows, n_cols),
     raft::make_device_matrix_view<value_t, int64_t>(distances.data(), n_left_rows, n_right_rows),

--- a/cpp/tests/stats/silhouette_score.cu
+++ b/cpp/tests/stats/silhouette_score.cu
@@ -1,5 +1,5 @@
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION.
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 #include "../test_utils.cuh"

--- a/cpp/tests/stats/silhouette_score.cu
+++ b/cpp/tests/stats/silhouette_score.cu
@@ -1,5 +1,5 @@
 /*
- * SPDX-FileCopyrightText: Copyright (c) 2024, NVIDIA CORPORATION.
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION.
  * SPDX-License-Identifier: Apache-2.0
  */
 #include "../test_utils.cuh"
@@ -7,14 +7,18 @@
 #include <cuvs/distance/distance.hpp>
 #include <cuvs/stats/silhouette_score.hpp>
 #include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/cuda_stream_pool.hpp>
 #include <raft/util/cudart_utils.hpp>
 
+#include <rmm/cuda_stream_pool.hpp>
 #include <rmm/device_uvector.hpp>
 
 #include <gtest/gtest.h>
 
 #include <algorithm>
+#include <cmath>
 #include <iostream>
+#include <memory>
 #include <random>
 
 namespace cuvs {
@@ -219,6 +223,43 @@ TEST_P(silhouetteScoreTestClass, Result)
   ASSERT_NEAR(batchedSilhouetteScore, truthSilhouetteScore, params.tolerance);
 }
 INSTANTIATE_TEST_CASE_P(silhouetteScore, silhouetteScoreTestClass, ::testing::ValuesIn(inputs));
+
+TEST(silhouetteScore, BatchedStreamPoolOrdering)
+{
+  constexpr int64_t n_rows = 4096;
+  constexpr int64_t n_cols = 2;
+  constexpr int n_labels   = 2;
+
+  std::vector<float> X(n_rows * n_cols);
+  std::vector<int> labels(n_rows);
+  for (int64_t i = 0; i < n_rows; ++i) {
+    X[2 * i]     = std::sin(0.01f * i);
+    X[2 * i + 1] = std::cos(0.013f * i);
+    labels[i]    = i % n_labels;
+  }
+
+  raft::resources handle;
+  raft::resource::set_cuda_stream_pool(handle, std::make_shared<rmm::cuda_stream_pool>(4));
+  auto stream = raft::resource::get_cuda_stream(handle);
+
+  rmm::device_uvector<float> d_X(X.size(), stream);
+  rmm::device_uvector<int> d_labels(labels.size(), stream);
+  raft::update_device(d_X.data(), X.data(), X.size(), stream);
+  raft::update_device(d_labels.data(), labels.data(), labels.size(), stream);
+
+  auto X_view = raft::make_device_matrix_view<const float, int64_t>(d_X.data(), n_rows, n_cols);
+  auto labels_view = raft::make_device_vector_view<const int, int64_t>(d_labels.data(), n_rows);
+  constexpr auto metric = cuvs::distance::DistanceType::L2SqrtUnexpanded;
+
+  auto expected =
+    cuvs::stats::silhouette_score(handle, X_view, labels_view, std::nullopt, n_labels, metric);
+
+  for (int repeat = 0; repeat < 8; ++repeat) {
+    auto actual = cuvs::stats::silhouette_score_batched(
+      handle, X_view, labels_view, std::nullopt, n_labels, n_rows, metric);
+    ASSERT_NEAR(actual, expected, 1e-4f);
+  }
+}
 
 }  // end namespace stats
 }  // end namespace cuvs

